sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -129,3 +135,12 @@ case class UserDefinedFunction protected[sql] (
     }
   }
 }
+
+object UserDefinedFunction {
+  // This is to keep backward compatibility for this case class.
+  // TODO: revisit this case class in Spark 3.0, and narrow down the public surface.
+  def unapply(obj: Any): Option[(AnyRef, DataType, Option[Seq[DataType]])] = obj match {
```

Comment:
Wenchen, can we have a small test to check if the previous pattern matching works or not, and can you double check this by resembling the case by `javap -v` manually at least since this looks not being checked?